### ISSUE TITLE
401 - Signature is incorrect for LinkedIn API requests

### ISSUE THREAD

**Issue by danieljsinclair** [DotNetOpenAuth/DotNetOpenAuth#175]
I thought this was due to parameter ordering but it isn't. Somewhere DotNetOpenAuth.AspNet is generates an invalid signature for the LinkedInClient API calls. It could be related to the way the URL is encoded, can't be sure. My requests are working both on the oauth-test-console and the LinkedInAuth sample, but DotNetOpenAuth.Asp.Net always generates a different (invalid) sig.

**Comment by danieljsinclair** [comment]
I've been debugging this, and it now looks like it might be related to the AuthenticationOnlyOAuthTokenManager class that the LinkedInClient uses by default. As it's name suggests, it's not providing an tokenSecret for the API call following authentication. This causes the signature to fail.

Presumably, it _should_ be doing;

```
throw new ProtocolException(OAuthStrings.ConsumerOrTokenSecretNotFound, ex);
```

from 

```
SignatureCallback(ITamperResistantProtocolMessage message)
```

but it doesn't. It just goes ahead and signs it with the wrong key as far as I can tell, causing a 401.

**Comment by rroa** [comment]
The problem is indeed with the token. The problem has todo with persistence, apparently when the OAuthClient method VerifyAuthenticationCore gets invoked the Token is empty. Here's an implementation of the library where the issue arises as well http://scatteredcode.wordpress.com/2011/12/01/dotnetopenauth-oauth-and-mvc-for-dummies/.

I have a custom client that fixes the issue by just avoiding to use the included client default constructor overload.

http://blog.mrroa.com/post/30454808112/asp-net-custom-linkedin-oauth-provider

**Comment by donwibier** [comment]
rroa, great job !!

I knew that it had something todo with the verifier but because of the well documented DotNetOpenAuth library, I couldn't find the proper way of getting the verifier in there!

**Comment by AnalogJ** [comment]
The code that was posted here didnt work out for me exactly, because when using DotNetOpenAuth.AspNet in a web app, the roundtrip back to the server means that the InMemoryTokenManager may be flushed. I was able to get everything working by using a temporary InMemoryTokenManager. I was also able to solve the persistence of the AccessToken secret by creating a Cookie backed Persistent token manger. I've provided full source code  on this [coderwall post](http://coderwall.com/p/rznlfg)
